File: test/test_rdf.cc
#include "edm4hep/utils/dataframe.h"

#include "edm4hep/MCParticleData.h"
#include "edm4hep/SimTrackerHitData.h"
#include "edm4hep/TrackerHitPlaneData.h"

#include <ROOT/RDataFrame.hxx>

#include <iostream>

int main(int, char*[]) {

  ROOT::EnableImplicitMT();

  std::cout << "Create RDataFrame ..." << std::endl;

  ROOT::RDataFrame df("events", "edm4hep_events.root");

  std::cout << "Apply selectors and define new branches ..." << std::endl;
  auto df2 = df.Define("MCParticles_pt", edm4hep::utils::pt<edm4hep::MCParticleData>, {"MCParticles"})
                 .Define("MCParticles_eta", edm4hep::utils::eta<edm4hep::MCParticleData>, {"MCParticles"})
                 .Define("MCParticles_cosTheta", edm4hep::utils::cos_theta<edm4hep::MCParticleData>, {"MCParticles"})
                 .Define("SimTrackerHits_r", edm4hep::utils::r<edm4hep::SimTrackerHitData>, {"SimTrackerHits"})
                 .Define("SimTrackerHit_pt", edm4hep::utils::pt<edm4hep::SimTrackerHitData>, {"SimTrackerHits"})
                 .Define("TrackerHits_r", edm4hep::utils::r<edm4hep::TrackerHitPlaneData>, {"TrackerHitPlanes"});

  std::string outfilename = "edm4hep_events_rdf.root";
  std::cout << "Writing snapshot to disk ... \t" << outfilename << std::endl;

  df2.Snapshot("events", outfilename,
               {"MCParticles_pt", "MCParticles_eta", "MCParticles_cosTheta", "SimTrackerHits_r", "SimTrackerHit_pt",
                "TrackerHits_r"});

  return 0;
}

File: test/test_rdf.py
#!/usr/bin/env python3

import ROOT
import edm4hep  # noqa: F401

ROOT.EnableImplicitMT()

print('Create RDataFrame ...')
df = ROOT.RDataFrame('events', 'edm4hep_events.root')

print('Apply selectors and define new branches ...')
df2 = (df.Define('MCParticles_pt', 'edm4hep::utils::pt(MCParticles)')
       .Define('MCParticles_eta', 'edm4hep::utils::eta(MCParticles)')
       .Define('MCParticles_cosTheta', 'edm4hep::utils::cos_theta(MCParticles)')
       .Define('SimTrackerHits_r', 'edm4hep::utils::r(SimTrackerHits)')
       .Define('SimTrackerHits_pt', 'edm4hep::utils::pt(SimTrackerHits)')
       .Define('TrackerHits_r', 'edm4hep::utils::r(TrackerHitPlanes)')
       )

filename = 'edm4hep_events_py_rdf.root'
print(f'Writing snapshot to disk ... \t{filename}')

df2.Snapshot('events', filename,
             ['MCParticles_pt',
              'MCParticles_eta',
              'MCParticles_cosTheta',
              'SimTrackerHits_r',
              'SimTrackerHits_pt',
              'TrackerHits_r'])
